### PR TITLE
feat: RP<>workspace mapping

### PR DESCRIPTION
**Pull request by eecsliu**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/db/postgres_rp_workspace_bindings.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+) ([]*RPWorkspaceBinding, *apiv1.Pagination, error) {
+	poolExists := false
+	for _, validPool := range resourcePools {
+		if validPool.PoolName == poolName {
+			poolExists = true
+			break
+		}
+	}
+	if !poolExists {
+		return nil, nil, errors.Errorf("pool with name %v doesn't exist in config", poolName)
```
Comment (maxrussell) [inline master/internal/db/postgres_rp_workspace_bindings.go:95]: 404 Error

File: master/internal/db/postgres_rp_workspace_bindings.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+				return rpWorkspaceBindings, pagination, nil
+			}
+			return nil, nil, err
+		}
+	}
+
+	return rpWorkspaceBindings, pagination, nil
+}
+
+// TODO find a good house for this function.
```
Comment (maxrussell) [inline master/internal/db/postgres_rp_workspace_bindings.go:123]: What's the decision about where this should go? If it's here, we should remove the TODO.

File: master/internal/db/postgres_test_utils.go [hunk 3]
```diff
@@ -323,3 +325,47 @@ func (db *PgDB) MustExec(t *testing.T, sql string, args ...any) sql.Result {
 	require.NoError(t, err, "failed to run query")
 	return out
 }
+
+// MockWorkspaces creates as many new workspaces as in workspaceNames and
+// returns a list of workspaceIDs.
```
Comment (maxrussell) [inline master/internal/db/postgres_test_utils.go:330]:
```suggestion
// MockWorkspaces creates as many new workspaces as in workspaceNames and
// returns their ids.
```

File: master/internal/experiment.go [hunk 4]
```diff
@@ -923,10 +937,16 @@ func (e *experiment) setWeight(ctx *actor.Context, weight float64) error {
 }
 
 func (e *experiment) setRP(ctx *actor.Context, msg sproto.SetResourcePool) error {
+	// TODO: secure this to prevent people from just changing RPs to whatever they want
```
Comment (maxrussell) [inline master/internal/experiment.go:940]: If this is done, we should remove this comment. If not, we should secure before release.

File: master/internal/workspace/postgres_workspace.go [hunk 5]
```diff
@@ -17,6 +22,45 @@ func WorkspaceByName(ctx context.Context, workspaceName string) (*model.Workspac
...
+			namesFound.Insert(workspace.Name)
+		}
+
+		for _, name := range workspaceNames {
+			if !namesFound.Contains(name) {
+				missing = append(missing, name)
+			}
+		}
+
+		return nil, fmt.Errorf("the following workspaces do not exist: %s", missing)
```
Comment (maxrussell) [inline master/internal/workspace/postgres_workspace.go:54]: 404 error

File: master/static/migrations/20230725094922_add-rp-workspace-bindings-constraint.tx.down.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE rp_workspace_bindings
+    DROP CONSTRAINT rp_workspace_pkey;
```
Comment (maxrussell) [inline master/static/migrations/20230725094922_add-rp-workspace-bindings-constraint.tx.down.sql:2]: Nit: all lines should have the POSIX-compliant terminating `\n`, including the final lines of files.

File: master/static/migrations/20230725094922_add-rp-workspace-bindings-constraint.tx.up.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE rp_workspace_bindings
+    ADD CONSTRAINT rp_workspace_pkey PRIMARY KEY (workspace_id, pool_name);
```
Comment (maxrussell) [inline master/static/migrations/20230725094922_add-rp-workspace-bindings-constraint.tx.up.sql:2]: Nit: all lines should have the POSIX-compliant terminating `\n`, including the final lines of files.

**Review by maxrussell — APPROVED** [review]
None of my comments seem like release blockers to me (as long as that security thing has been done). So I'm comfortable with this being released before things like nits and cleanup get done.

File: master/internal/db/postgres_rp_workspace_bindings.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+	poolExists := false
+	for _, pool := range resourcePools {
+		if poolName == pool.PoolName {
+			poolExists = true
+			break
+		}
+	}
+
+	if !poolExists {
+		return errors.Errorf("pool with name %v doesn't exist in config", poolName)
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/db/postgres_rp_workspace_bindings.go:43]: for posterity: fix this and the above 500 error

File: master/internal/db/postgres_rp_workspace_bindings.go [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+	workspaceIds []int32, poolName string,
+) error {
+	if len(workspaceIds) == 0 {
+		return nil
+	}
+	// throw error if any of bindings don't exist
+	for _, workspaceID := range workspaceIds {
+		var rpWorkspaceBindings []*RPWorkspaceBinding
+		exists, err := Bun().NewSelect().Model(&rpWorkspaceBindings).Where("pool_name = ?",
+			poolName).Where("workspace_id = ?", workspaceID).Exists(ctx)
```
[the remark this answers is not in the document]
Comment (eecsliu) [reply master/internal/db/postgres_rp_workspace_bindings.go]: done and dusted!

File: master/internal/db/postgres_rp_workspace_bindings.go (hunk 2 above)
Comment (eecsliu) [reply master/internal/db/postgres_rp_workspace_bindings.go:133]: There hasn't been a decision yet - the original function is pretty far removed, so this is probably that should be tech debt

File: master/internal/experiment.go (hunk 4 above)
Comment (eecsliu) [reply master/internal/experiment.go]: this comment shouldn't exist anymore...I'll fix if it still sows up later

File: harness/determined/common/experimental/resource_pool.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+        def get_with_offset(offset: int) -> bindings.v1ListWorkspacesBoundToRPResponse:
+            return bindings.get_ListWorkspacesBoundToRP(
+                session=self._session,
+                offset=offset,
+                resourcePoolName=self.name,
+            )
+
+        resps = api.read_paginated(get_with_offset)
+        workspace_names = [
```
Comment (wes-turner) [inline harness/determined/common/experimental/resource_pool.py:83]: Nested `for`s are hard to unwind in list comprehensions (and are contraindicated by [the style guide](https://google.github.io/styleguide/pyguide.html#274-decision) we're trying to use).

(this, BTW, is more suggestion than requirement. Compliance with the style guide is still aspirational)

One approach is to ditch the comprehension for a `for` loop with accumulator. Another is to continue to use a comprehension but just find another way to break up the nesting. `itertools.chain.from_iterable` is a standard tool for that.

I'm sorry if I missed this before or gave conflicting advice.

File: harness/tests/determined/common/experimental/test_workspace.py [hunk 11]
```diff
@@ -71,3 +87,21 @@ def test_workspace_constructor_populates_id_from_name(
 def test_workspace_constructor_doesnt_populate_name_from_id(standard_session: api.Session) -> None:
     ws = workspace.Workspace(session=standard_session, workspace_id=1)
     assert ws.name is None
+
+
+@responses.activate
+def test_list_bounded_resource_pools(
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_workspace.py:93]: This function name is hard for me -- I don't know what it's supposed to test. Could you please rename in the style suggested at https://hpe-aiatscale.atlassian.net/wiki/spaces/ENGINEERIN/pages/1351909582/Python+Testing#Best-practices?

File: harness/tests/determined/common/experimental/test_workspace.py [hunk 12]
```diff
@@ -23,6 +23,22 @@ def single_item_workspaces() -> bindings.v1GetWorkspacesResponse:
     )
 
 
+@pytest.fixture
```
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_workspace.py:26]: I think a summary, 1-line docstring would be helpful here.

**Pull request merged**